cmd/agent.go

```diff
@@ -214,6 +217,55 @@ func (ar AgentRunner) getRunnerInstance(logger hclog.Logger, path string) (runne
 	return runnerInstance, nil
 }
 
-func (ar AgentRunner) closePluginClients() {
+// DownloadPlugin checks whether we need to download the source plugin, or whether it is already on the file system.
+// If it isn't on the filesystem, we'll download it and return the final destination for use.
+func (ar *AgentRunner) DownloadPlugin(source string) (usablePlugin string, err error) {
+	// First we check if the source is a path that exists on the fs.
+	// If it does exist, it means we've been passed a binary, and we can just use it as is.
+	_, err = os.ReadFile(source)
+
+	if err == nil {
+		// The file exists. Just return it.
+		ar.logger.Debug("Found plugin locally. Using local binary.", "Binary", source)
+		return source, err
+	}
+
+	if !os.IsNotExist(err) {
+		// The error we've received is something other than not exists.
+		// Exit early with the error
+		return "", err
+	}
+
+	if internal.IsOCI(source) {
+		ar.logger.Debug("Plugin looks like an OCI endpoint. Attempting to download.", "Source", source)
+		tag, err := name.NewTag(source)
+		if err != nil {
+			return "", err
+		}
+
+		destination := path.Join(AgentPluginDir, tag.RepositoryStr(), tag.Identifier())
+
+		downloaderImpl, err := oci.NewDownloader(
+			tag,
+			destination,
+		)
+		if err != nil {
+			return "", err
+		}
+		err = downloaderImpl.Download()
+		if err != nil {
+			return "", err
+		}
+		pluginBinary := path.Join(destination, "plugin")
```

Comment:
Given that the name `plugin` is hardcoded here (which is not ideal but can live with for now) it might be better if this was abstracted out.